arwen/contexts/async_synchronous.go

```diff
@@ -0,0 +1,243 @@
+package contexts
+
+import (
+	"math/big"
+
+	"github.com/ElrondNetwork/arwen-wasm-vm/arwen"
+	"github.com/ElrondNetwork/elrond-go/core/vmcommon"
+)
+
+func (context *asyncContext) executeSynchronousCalls() error {
+	for groupIndex, group := range context.AsyncCallGroups {
+		for _, call := range group.AsyncCalls {
+			if call.ExecutionMode != arwen.SyncExecution {
+				continue
+			}
+
+			err := context.executeSyncCall(call)
+			if err != nil {
+				return err
+			}
+		}
+
+		group.DeleteCompletedAsyncCalls()
+
+		// If all the AsyncCalls in the AsyncCallGroup were executed synchronously,
+		// then the AsyncCallGroup can have its callback executed.
+		if group.IsCompleted() {
+			context.executeCallGroupCallback(group)
+			context.DeleteCallGroup(groupIndex)
+		}
+	}
+	return nil
+}
+
+func (context *asyncContext) executeSyncCall(asyncCall *arwen.AsyncCall) error {
+	destinationCallInput, err := context.createSyncCallInput(asyncCall)
+	if err != nil {
+		return err
+	}
+
+	vmOutput, err := context.host.ExecuteOnDestContext(destinationCallInput)
+
+	// The vmOutput instance returned by host.ExecuteOnDestContext() is never nil,
+	// by design. Using it without checking for err is safe here.
+	asyncCall.UpdateStatus(vmOutput.ReturnCode)
+
+	callbackVMOutput, callbackErr := context.executeSyncCallback(asyncCall, vmOutput, err)
+	context.finishSyncExecution(callbackVMOutput, callbackErr)
+
+	return nil
+}
+
+func (context *asyncContext) executeSyncCallback(
+	asyncCall *arwen.AsyncCall,
+	vmOutput *vmcommon.VMOutput,
+	err error,
+) (*vmcommon.VMOutput, error) {
+
+	callbackInput, err := context.createSyncCallbackInput(asyncCall, vmOutput, err)
+	if err != nil {
+		return nil, err
+	}
+
+	return context.host.ExecuteOnDestContext(callbackInput)
+}
+
+// executeCallGroupCallback synchronously executes the designated callback of
+// the AsyncCallGroup, as it was set with SetGroupCallback().
+//
+// Gas for the execution has been already paid for when SetGroupCallback() was
+// set. The remaining gas is refunded to context.CallerAddr, which initiated
+// the call and paid for the gas in the first place.
+func (context *asyncContext) executeCallGroupCallback(group *arwen.AsyncCallGroup) {
+	if !group.HasCallback() {
+		return
+	}
+
+	input := context.createGroupCallbackInput(group)
+	vmOutput, err := context.host.ExecuteOnDestContext(input)
+	context.finishSyncExecution(vmOutput, err)
+}
+
+func (context *asyncContext) finishSyncExecution(vmOutput *vmcommon.VMOutput, err error) {
+	if err == nil {
```

Comment:
is this correct ? shouldn't be like err != nil, vmOutput.ReturnCode != vmcommon.Ok for error checking?

Reply:
Yes, `finishSyncExecution()` is intended to do nothing for `vmcommon.Ok`, but it will add extra error-related information to the current output if `vmOutput.ReturnCode` is not `vmcommon.Ok`.

Either way, this method will be revised, because it currently returns a `vmOutput` and an `error`, but these are never used by the code that call it. Added TODO.